qiskit_nature/algorithms/initial_points/mp2_initial_point.py

```diff
@@ -0,0 +1,302 @@
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""The MP2InitialPoint class to compute an initial point for the VQE Ansatz parameters."""
+
+from __future__ import annotations
+
+import numpy as np
+from qiskit_nature.exceptions import QiskitNatureError
+
+from qiskit_nature.properties.second_quantization.electronic import ElectronicEnergy
+from qiskit_nature.properties.second_quantization.electronic.bases import ElectronicBasis
+from qiskit_nature.circuit.library import UCC
+from qiskit_nature.properties.second_quantization.second_quantized_property import (
+    GroupedSecondQuantizedProperty,
+)
+
+from .initial_point import InitialPoint
+
+
+class MP2InitialPoint(InitialPoint):
+    """Moller-Plesset 2nd order initial point generator.
+
+    A class for computing the Moller-Plesset second order (MP2) corrections to use as an initial
+    point with a VQE algorithm alongside a UCC ansatz.
+
+    The MP2 calculation requires the two-body molecular orbital matrix, the orbital energies,
+    and the Hartree-Fock reference energy from the ``ElectronicEnergy`` in the grouped
+    property.
+
+    An MP2InitialPoint uses the same excitations as generated by the UCC ansatz to ensure that
+    the coefficients are mapped correctly in the initial point. Only initial point values that
+    correspond to double-excitations will be non-zero.
+
+    The coefficients and energy deltas are computed using the ``get_initial_point`` method.
+    Thereafter, they can be accessed via the ``initial_point`` and ``energy_deltas`` properties.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.threshold = 1e-12
+        self._corrections: dict[str, tuple[float, float]] = None
+
+    @property
+    def grouped_property(self) -> GroupedSecondQuantizedProperty:
+        """The grouped property."""
+        return self._grouped_property
+
+    @grouped_property.setter
+    def grouped_property(self, grouped_property: GroupedSecondQuantizedProperty) -> None:
+        """The grouped property.
+
+        The grouped property is required to contain the ElectronicEnergy, which must contain
+        the two-body molecular orbital matrix, the orbital energies, and the Hartree-Fock
+        reference energy.
+
+        Raises:
+            QiskitNatureError: If the ``ElectronicEnergy`` is missing or the two-body molecular
+            orbital matrix, the orbital energies, and/or the Hartree-Fock reference energy are
+            missing from ``ElectronicEnergy``.
+        """
+        try:
+            electronic_energy: ElectronicEnergy = grouped_property.get_property(ElectronicEnergy)
+        except TypeError as invalid_grouped_property:
+            raise QiskitNatureError from invalid_grouped_property
```

Comment:
I do not think this will ever happen because `get_property` will safely return `None` even if `ElectronicEnergy` does not exist.